packages/projection-typeorm/src/createSinks.ts

```diff
@@ -0,0 +1,169 @@
+import 'reflect-metadata';
+import * as supportedSinks from './sinks';
+import { BlockEntity } from './entity/Block.entity';
+import { ChainSyncEventType } from '@cardano-sdk/core';
+import { DataSource } from 'typeorm';
+import { Logger } from 'ts-log';
+import {
+  NEVER,
+  Observable,
+  ReplaySubject,
+  Subject,
+  concat,
+  concatMap,
+  finalize,
+  from,
+  map,
+  share,
+  switchMap,
+  tap,
+  timer
+} from 'rxjs';
+import {
+  Operators,
+  ProjectionSinks,
+  Projections,
+  Sinks,
+  UnifiedProjectorEvent,
+  UnifiedProjectorObservable
+} from '@cardano-sdk/projection';
+import { RetryBackoffConfig, retryBackoff } from 'backoff-rxjs';
+import { TypeormStabilityWindowBuffer, TypeormStabilityWindowBufferProps } from './TypeormStabilityWindowBuffer';
+import { WithLogger } from '@cardano-sdk/util';
+import { WithTypeormContext } from './types';
+import { finalizeWithLatest } from '@cardano-sdk/util-rxjs';
+import { isRecoverableTypeormError } from './util';
+import omit from 'lodash/omit';
+
+export type ReconnectionConfig = Omit<RetryBackoffConfig, 'shouldRetry'>;
+const defaultReconnectionConfig: ReconnectionConfig = { initialInterval: 10, maxInterval: 5000 };
+
+export interface TypeormSinksProps extends TypeormStabilityWindowBufferProps, WithLogger {
+  /**
+   * Re-subscribes to dataSource$ on connection error
+   */
+  dataSource$: Observable<DataSource>;
+  reconnectionConfig?: ReconnectionConfig;
+}
+
+const TypeormContextProps: Array<keyof WithTypeormContext> = ['blockEntity', 'queryRunner', 'transactionCommit$'];
+
+export type SupportedProjections = Pick<Projections.AllProjections, keyof typeof supportedSinks>;
+
+const withQueryRunner = (dataSource: DataSource, buffer: TypeormStabilityWindowBuffer, logger: Logger) =>
+  concat(
+    from(
+      (async () => {
+        const queryRunner = dataSource.createQueryRunner('master');
+        await queryRunner.connect();
+        // buffer has to emit tip$ and tail$ for projectIntoSink
+        // to find intersection before emitting any events
+        await buffer.initialize(queryRunner);
+        return { queryRunner };
+      })()
+    ),
+    NEVER
+  ).pipe(
+    finalizeWithLatest(async (evt) => {
```

Comment:
I don't think this will ever finalize due to `NEVER` being the last observable in the concat pipe 🤔

Reply:
It does, when unsubscribed by the subscriber